File: check_files.py
import os
import subprocess
import gc

# Function to run Dafny verification on a list of files
def run_dafny_verification(file_paths, start_index):
    failed_files = []
    try:
        for index, file_path in enumerate(file_paths, start=start_index):
            # Run Dafny verification command
            result = subprocess.run(["dafny", "verify", file_path, "--verification-time-limit", "60", "--cores", "7"], capture_output=True, text=True)
            # Check if Dafny verification was successful
            if result.returncode != 0:
                failed_files.append((file_path, result.stderr))
            print(f"Verification attempt {index} for file: {file_path}")
            # Clear memory
            gc.collect()
    except Exception as e:
        print(f"Error occurred: {str(e)}")
    return failed_files

def main():
    # Get the directory of the current Python script
    script_directory = os.path.dirname(os.path.abspath(__file__))
    # Directory containing Dafny files
    directory = os.path.join(script_directory, "compilable")

    # List to store file paths
    file_paths = []

    # Iterate through files in the directory and sort alphabetically
    for filename in sorted(os.listdir(directory)):
        if filename.endswith(".dfy"):
            file_path = os.path.join(directory, filename)
            file_paths.append(file_path)

    # Break files into chunks of 100
    chunk_size = 100
    file_chunks = [file_paths[i:i+chunk_size] for i in range(0, len(file_paths), chunk_size)]

    # Process each chunk sequentially
    start_index = 1
    for chunk_index, chunk in enumerate(file_chunks):
        print(f"Processing chunk {chunk_index + 1}...")
        failed_files = run_dafny_verification(chunk, start_index)

        # Write failed verifications to output file
        with open(f"failed_files_chunk_{chunk_index + 1}.txt", "w") as output_file:
            for failed_file, error_message in failed_files:
                output_file.write(f"File '{failed_file}' failed verification with error: {error_message}\n")
        
        start_index += len(chunk)

if __name__ == "__main__":
    main()

File: check_files.rs
use std::process::{Command, Stdio};
use std::fs;
use std::io::Write;

// Function to run Dafny verification on a list of files
fn run_dafny_verification(file_paths: Vec<String>, start_index: usize) -> Vec<(String, String)> {
    let mut failed_files = vec![];
    for (index, file_path) in file_paths.iter().enumerate() {
        let mut command = Command::new("dafny");
        command.args(&["verify", file_path, "--verification-time-limit", "60", "--cores", "7"]);
        command.stdout(Stdio::piped());
        command.stderr(Stdio::piped());
        
        match command.output() {
            Ok(output) => {
                if !output.status.success() {
                    let error_message = String::from_utf8_lossy(&output.stderr).to_string();
                    failed_files.push((file_path.clone(), error_message));
                }
                println!("Verification attempt {} for file: {}", start_index + index + 1, file_path);
            },
            Err(err) => {
                println!("Error occurred: {}", err);
            }
        }
        
        // Clear memory
        // No equivalent for gc.collect() in Rust, memory management is handled automatically
    }
    failed_files
}

fn main() {
    // Directory containing Dafny files
    let directory = "./compilable";

    // List to store file paths
    let mut file_paths = Vec::new();

    // Iterate through files in the directory and sort alphabetically
    if let Ok(entries) = fs::read_dir(directory) {
        for entry in entries {
            if let Ok(entry) = entry {
                if let Some(file_name) = entry.file_name().to_str() {
                    if file_name.ends_with(".dfy") {
                        file_paths.push(entry.path().to_str().unwrap().to_string());
                    }
                }
            }
        }
    }
    file_paths.sort(); // Sort alphabetically

    // Break files into chunks of 100
    let chunk_size = 100;
    let file_chunks: Vec<_> = file_paths.chunks(chunk_size).collect();

    // Process each chunk sequentially
    let mut start_index = 1;
    for (chunk_index, chunk) in file_chunks.iter().enumerate() {
        println!("Processing chunk {}...", chunk_index + 1);
        let failed_files = run_dafny_verification(chunk.to_vec(), start_index);

        // Write failed verifications to output file
        let output_file_path = format!("failed_files_chunk_{}.txt", chunk_index + 1);
        let mut output_file = fs::File::create(output_file_path).expect("Failed to create output file");
        for (failed_file, error_message) in failed_files {
            writeln!(&mut output_file, "File '{}' failed verification with error: {}", failed_file, error_message).expect("Failed to write to output file");
        }
        
        start_index += chunk.len();
    }
}
